Expose environment variables to client

amoore108 opened the pull request:
Overall changes
======
- To enable the user of per-environment `process.env` variables, we need to expose them to the client so that the application can then read them in and make use of them. This essentially "bakes in" the environment variables so that they are always correct for the environment the application is running in.
- This requires exposing a `<script>` tag in the `head` that store environment variables for both the Express and Next.js apps. These values are set on the server-side in the `document` and `_document` files in the respectful apps. They are then read in on the client-side from the `window.SIMORGH_ENV_VARS` object and from the general `process.env` on the server-side.
- A new utility called `getEnvConfig()` has been added and replaces instances of `process.env` across the app where appropriate. It is used like `getEnvConfig().SIMORGH_APP_ENV`. This ensures that the correct value is used on **both** server and client renders, as it checks which object to read the values from based on whether or not its on the server or client.

Testing
======
This will require a near full regression test of the application to ensure at least:
1. Images are served from the correct location (`ichef.test` on the preview and test environments and `ichef.` on the production environment)
2. ATI is sending to the correct location (`https://logws1363.ati-host.net?` on preview and test and `https://a1.api.bbc.co.uk/hit.xiti?` on production)
3. Generally that the environment is talking to the correct environment. For example the BFF is talking to the right environment based on whether or not the request was made on preview/test or production 

Helpful Links
======
_Add Links to useful resources related to this PR if applicable._

[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

amoore108 committed:
Testing exposing clientside env vars

amoore108 committed:
Test in Express app

amoore108 committed:
Merge branch 'latest' into nextjs-expose-client-env-vars

amoore108 committed:
import fix

amoore108 committed:
Rename env variables object to be clearer and what it does

amoore108 committed:
Remove `SIMORGH_APP_ENV` from `next.config.js`

amoore108 committed:
Merge branch 'latest' into nextjs-expose-client-env-vars

amoore108 committed:
Use initialiser method to allow `process.env` overrides

- This is mainly to support tests that set the `process.env` values within the tests

amoore108 committed:
Update index.ts

amoore108 committed:
Test with `SIMORGH_ATI_BASE_URL`

amoore108 committed:
Type check `SIMORGH_APP_ENV`

amoore108 committed:
Update _document.page.tsx

amoore108 committed:
Update index.ts

amoore108 committed:
Merge branch 'latest' into nextjs-expose-client-env-vars

amoore108 committed:
Update component.jsx

amoore108 committed:
Update `process.env` references to use `getEnvConfig()`

amoore108 committed:
Add missing cache config

amoore108 committed:
Set ichef default back

amoore108 committed:
Fix import

amoore108 committed:
Merge branch 'latest' into nextjs-expose-client-env-vars

amoore108 committed:
Update index.ts

amoore108 committed:
Snapshot update

amoore108 committed:
Test removing `getEnvConfig` from `getOriginContext`

amoore108 committed:
Merge branch 'latest' into nextjs-expose-client-env-vars

amoore108 committed:
Revert "Test removing `getEnvConfig` from `getOriginContext`"

This reverts commit 81e26dba2c8cea903e9fe664ff897aaa862bf5df.

amoore108 committed:
Update index.ts

amoore108 committed:
Test removing path aliasing

amoore108 committed:
Merge branch 'latest' into nextjs-expose-client-env-vars

amoore108 committed:
Add test

andrewscfc commented on src/server/Document/component.jsx:
```diff
@@ -84,25 +87,34 @@ const Document = ({
             dangerouslySetInnerHTML={{
-              __html: `window.SIMORGH_DATA=${serialisedData}`,
+              // Read env variables from the server and expose them to the client
+              __html: `window.SIMORGH_ENV_VARS=${appEnvVariables}`,
```
I think you might be able to avoid the `dangerouslySetInnerHTML` which the approach in this PR description: https://github.com/bbc/simorgh/pull/6838 

We have a ticket to try this for Simorgh Data too

amoore108 replied:
Could follow that approach, although it doesn't seem to get rid of `dangerouslySetInnerHTML`? Looks like it just reads the text content rather than the `window`.

amoore108 committed:
Improve tests

amoore108 committed:
Update index.test.ts

amoore108 committed:
Remove `process.env` from clientside test

- These should be coming from the `window` object, so its more accurate to remove these

karinathomasbbc reviewed (approved):
So, it looks like pretty much all occurences of `process.env` have been removed from the codebase, and hopefully if we need a different environment variable somewhere, then we'll copy and paste the pattern we see i.e. use `getEnvConfig()` instead of `process.env`

I wonder if there's something we could do in terms of an automated test or linting to ensure that folk don't accidentally use `process.env`? Not sure if this is possible, just a thought, because it will be tricky to remember in future, or relies on code reviewers remembering that we should use `getEnvConfig()` instead of `process.env`...

karinathomasbbc commented on src/app/lib/utilities/getEnvConfig/index.ts:
```diff
@@ -0,0 +1,46 @@
+export type EnvConfig = ReturnType<typeof getProcessEnvAppVariables>;
+
+// Any environment variables added here will be available to the client and server
+export const getProcessEnvAppVariables = () => ({
```
Per my comment above, is there some way we could ensure that these environment variables match what's in our local/test/live environment files? Just to avoid issues if a dev adds a new environment variable but forgets to add it here. Could typescript help us somehow?

amoore108 replied:
We could convert the `.env` files to JSON objects and type check them? The only problem is this would only apply to the Express app, as the `.env` files for the Next.js app are set ultimately set and read from the Lambda function itself.

amoore108 commented:
> So, it looks like pretty much all occurences of `process.env` have been removed from the codebase, and hopefully if we need a different environment variable somewhere, then we'll copy and paste the pattern we see i.e. use `getEnvConfig()` instead of `process.env`
> 
> I wonder if there's something we could do in terms of an automated test or linting to ensure that folk don't accidentally use `process.env`? Not sure if this is possible, just a thought, because it will be tricky to remember in future, or relies on code reviewers remembering that we should use `getEnvConfig()` instead of `process.env`...

Yea its a tricky one to automate to be honest, as in some instances we should use `process.env`, for example in scripts outside of the built application, directly within server-only files like `server/index.js` (although you _probably_ could use the new utility) and in areas where the `process.env` is determined externally, like the `BFF_PATH`.

amoore108 committed:
Merge branch 'latest' into nextjs-expose-client-env-vars
